File: 88.Merge-Sorted-Array.go
// https://leetcode.com/problems/merge-sorted-array/
//
// algorithms
// Easy (34.21%)
// Total Accepted:    306,449
// Total Submissions: 895,894
// beats 100.0% of golang submissions

package leetcode

func merge(nums1 []int, m int, nums2 []int, n int) {
	for idx, target := range nums2 {
		insertIdx := binaryInsert(nums1, idx+m, target)
		for i := idx + m - 1; i >= insertIdx; i-- {
			nums1[i+1] = nums1[i]
		}
		nums1[insertIdx] = target
	}
}

func binaryInsert(nums1 []int, tail, target int) int {
	head, middle := 0, -1

	for head < tail {
		middle = head + (tail-head)/2
		if nums1[middle] > target {
			tail = middle
		} else if nums1[middle] < target {
			head = middle + 1
		} else {
			return middle
		}
	}

	return tail
}

File: 88.Merge-Sorted-Array.py
# https://leetcode.com/problems/gray-code/
#
# algorithms
# Easy (35.75%)
# Total Accepted:    365,649
# Total Submissions: 1,022,741
# beats 95.52% of python submissions


class Solution(object):
    def merge(self, nums1, m, nums2, n):
        """
        :type nums1: List[int]
        :type m: int
        :type nums2: List[int]
        :type n: int
        :rtype: void Do not return anything, modify nums1 in-place instead.
        """
        while m and n:
            if nums1[m - 1] > nums2[n - 1]:
                nums1[m + n - 1] = nums1[m - 1]
                m -= 1
            else:
                nums1[m + n - 1] = nums2[n - 1]
                n -= 1

        nums1[:n] = nums2[:n]

        return nums1
